`openldap-2.6.2` fails to build with `ld` input file not found

### Steps To Reproduce
Steps to reproduce the behavior:
1. build nixpkgs#openldap from 8626023077019a3b6309a5ebb2e58dfdf8cd178d

### Build log
```
openldap> cc -g -O2 -I../../include -I../../include       -c -o ldif-filter.o ldif-filter.c
openldap> cc -g -O2 -I../../include -I../../include       -c -o slapd-watcher.o slapd-watcher.c
openldap> /nix/store/x8ny3gm8j11gmy4lafcy524b2d3mzdyj-bash-5.1-p16/bin/bash ../../libtool --mode=link cc  -g -O2     -o ldif-filter ldif-filter.o slapd-common.o ../../libraries/libldap/libldap.la ../../libraries/liblutil/liblutil.a ../../libraries/libldap/libldap.la ../../libraries/liblber/liblber.la -lsasl2 -lssl -lcrypto  -lcrypt
openldap> libtool: link: cc -g -O2 -o .libs/ldif-filter ldif-filter.o slapd-common.o  ../../libraries/liblutil/liblutil.a ../../libraries/libldap/.libs/libldap.so /build/openldap-2.6.2/libraries/liblber/.libs/liblber.so ../../libraries/liblber/.libs/liblber.so /nix/store/n6n2lc5jqryhmp85nq9rr130kchp42v4-cyrus-sasl-2.1.28/lib/libsasl2.so -lssl -lcrypto -lcrypt -Wl,-rpath -Wl,/nix/store/hndyqjqqshss366zfcqq2y1nz658mnny-openldap-2.6.2/lib -Wl,-rpath -Wl,/nix/store/n6n2lc5jqryhmp85nq9rr130kchp42v4-cyrus-sasl-2.1.28/lib
openldap> /nix/store/rnis90bng6iv4l1z9g7crcsr42xghs10-binutils-2.38/bin/ld: cannot find slapd-common.o: No such file or directory
openldap> collect2: error: ld returned 1 exit status
openldap> make[2]: *** [Makefile:320: ldif-filter] Error 1
openldap> make[2]: Leaving directory '/build/openldap-2.6.2/tests/progs'
openldap> make[1]: *** [Makefile:394: all-common] Error 1
openldap> make[1]: Leaving directory '/build/openldap-2.6.2/tests'
openldap> make: *** [Makefile:320: all-common] Error 1
error: builder for '/nix/store/jvl2n4802hlpmksnkrn782fsb52zmfl4-openldap-2.6.2.drv' failed with exit code 2;
       last 10 log lines:
       > cc -g -O2 -I../../include -I../../include       -c -o slapd-watcher.o slapd-watcher.c
       > /nix/store/x8ny3gm8j11gmy4lafcy524b2d3mzdyj-bash-5.1-p16/bin/bash ../../libtool --mode=link cc  -g -O2     -o ldif-filter ldif-filter.o slapd-common.o ../../libraries/libldap/libldap.la ../../libraries/liblutil/liblutil.a ../../libraries/libldap/libldap.la ../../libraries/liblber/liblber.la -lsasl2 -lssl -lcrypto  -lcrypt
       > libtool: link: cc -g -O2 -o .libs/ldif-filter ldif-filter.o slapd-common.o  ../../libraries/liblutil/liblutil.a ../../libraries/libldap/.libs/libldap.so /build/openldap-2.6.2/libraries/liblber/.libs/liblber.so ../../libraries/liblber/.libs/liblber.so /nix/store/n6n2lc5jqryhmp85nq9rr130kchp42v4-cyrus-sasl-2.1.28/lib/libsasl2.so -lssl -lcrypto -lcrypt -Wl,-rpath -Wl,/nix/store/hndyqjqqshss366zfcqq2y1nz658mnny-openldap-2.6.2/lib -Wl,-rpath -Wl,/nix/store/n6n2lc5jqryhmp85nq9rr130kchp42v4-cyrus-sasl-2.1.28/lib
       > /nix/store/rnis90bng6iv4l1z9g7crcsr42xghs10-binutils-2.38/bin/ld: cannot find slapd-common.o: No such file or directory
       > collect2: error: ld returned 1 exit status
       > make[2]: *** [Makefile:320: ldif-filter] Error 1
       > make[2]: Leaving directory '/build/openldap-2.6.2/tests/progs'
       > make[1]: *** [Makefile:394: all-common] Error 1
       > make[1]: Leaving directory '/build/openldap-2.6.2/tests'
       > make: *** [Makefile:320: all-common] Error 1
       For full logs, run 'nix log /nix/store/jvl2n4802hlpmksnkrn782fsb52zmfl4-openldap-2.6.2.drv'.
error: 1 dependencies of derivation '/nix/store/mvd7nwaqzcs8v5md0bbz744bws9ijv7f-apr-util-1.6.1.drv' failed to build
error: 1 dependencies of derivation '/nix/store/23sc6aj5lcnj5r5pm2pd6i8wrlpr8bw5-gnupg-2.3.4.drv' failed to build
error: 1 dependencies of derivation '/nix/store/5ra0hgykvqxz13kpfprxavq4i6mf69yj-heimdal-7.7.0.drv' failed to build
```
[full log](https://gist.github.com/zseri/5b3fcb902e7344a81e398a87efb91f6a)

Naive guess: maybe some dependency is not correctly declared in the Makefile.

### Additional context
```console
[user@system:~]$ grep experim /etc/nix/nix.conf 
experimental-features = nix-command ca-derivations flakes
```

### Notify maintainers
@ajs124 @dasJ @mweinelt

### Metadata
```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.37, NixOS, 22.05 (Quokka), 22.05.20220506.ae4d683`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.8.0pre20220324_540d7e3`
 - nixpkgs: `/etc/nix-flake-channels/nixpkgs`
```

--- a/pkgs/development/libraries/openldap/default.nix
+++ b/pkgs/development/libraries/openldap/default.nix
@@ -1,6 +1,7 @@
 { lib
 , stdenv
 , fetchurl
+, fetchpatch
 
 # dependencies
 , cyrus_sasl
@@ -20,6 +21,39 @@ stdenv.mkDerivation rec {
     url = "https://www.openldap.org/software/download/OpenLDAP/openldap-release/${pname}-${version}.tgz";
     hash = "sha256-gdCTRSMutiSG7PWsrNLFbAxFtKbIwGZhLn9CGiOhz4c";
   };
+
+  patches = [
+    # ITS#9840 - ldif-filter: fix parallel build failure
+    (fetchpatch {
+      url = "https://github.com/openldap/openldap/commit/7d977f51e6dfa570a471d163b9e8255bdd3dc12f.patch";
+      hash = "sha256:1vid6pj2gmqywbghnd380x19ml241ldc1fyslb6br6q27zpgbdlp";
+    })
+    # ITS#9840 - libraries/Makefile.in: ignore the mkdir errors
+    (fetchpatch {
+      url = "https://github.com/openldap/openldap/commit/71f24015c312171c00ce94c9ff9b9c6664bdca8d.patch";
+      hash = "sha256:1a81vv6nkhgiadnj4g1wyzgzdp2zd151h0vkwvv9gzmqvhwcnc04";
+    })
+    # ITS#7165 back-mdb: check for stale readers on
+    (fetchpatch {
+      url = "https://github.com/openldap/openldap/commit/7e7f01c301db454e8c507999c77b55a1d97efc21.patch";
+      hash = "sha256:1fc2yck2gn3zlpfqjdn56ar206npi8cmb8yg5ny4lww0ygmyzdfz";
+    })
+    # ITS#9858 back-mdb: delay indexer task startup
+    (fetchpatch {
+      url = "https://github.com/openldap/openldap/commit/ac061c684cc79d64ab4089fe3020921a0064a307.patch";
+      hash = "sha256:01f0y50zlcj6n5mfkmb0di4p5vrlgn31zccx4a9k5m8vzxaqmw9d";
+    })
+    # ITS#9858 back-mdb: fix index reconfig
+    (fetchpatch {
+      url = "https://github.com/openldap/openldap/commit/c43c7a937cfb3a781f99b458b7ad71eb564a2bc2.patch";
+      hash = "sha256:02yh0c8cyx14iir5qhfam5shrg5d3115s2nv0pmqdj6najrqc5mm";
+    })
+    # ITS#9157: check for NULL ld
+    (fetchpatch {
+      url = "https://github.com/openldap/openldap/commit/6675535cd6ad01f9519ecd5d75061a74bdf095c7.patch";
+      hash = "sha256:0dali5ifcwba8400s065f0fizl9h44i0mzb06qgxhygff6yfrgif";
+    })
+  ];
 
   # TODO: separate "out" and "bin"
   outputs = [
@@ -59,7 +93,7 @@ stdenv.mkDerivation rec {
     "ac_cv_func_memcmp_working=yes"
   ] ++ lib.optional stdenv.isFreeBSD "--with-pic";
 
-  makeFlags= [
+  makeFlags = [
     "CC=${stdenv.cc.targetPrefix}cc"
     "STRIP="  # Disable install stripping as it breaks cross-compiling. We strip binaries anyway in fixupPhase.
     "prefix=${placeholder "out"}"
@@ -116,6 +150,6 @@ stdenv.mkDerivation rec {
     description = "An open source implementation of the Lightweight Directory Access Protocol";
     license = licenses.openldap;
     maintainers = with maintainers; [ ajs124 das_j hexa ];
-    platforms   = platforms.unix;
+    platforms = platforms.unix;
   };
 }

